### PR TITLE
Flink: flink read iceberg upsert data use streaming mode

### PR DESCRIPTION
**Pull request by hameizi**
Now, flink read iceberg use streaming mode ignore 'overwrite' snapshot, so user can't read the delete data real-time.
This PR first emit the eqdelete as delete rowdata(-U), and then emit the rowdata(+I) that join adddata and posdelete by function applyPosdelete. And then flink can keep one primary table from read iceberg primary table.

### CODE REVIEW

**Comment by hameizi** [comment]
@openinx @stevenzwu Could you help take a look?

**Comment by openinx** [comment]
Thanks for @hameizi for contribution, I will take a look  tomorrow.

**Comment by hameizi** [comment]
@openinx Could you help take a look again?

**Comment by xinbinhuang** [comment]
Hi @hameizi, thanks for the great work! I'm really interested in this patch to support `overwrite` snapshots because 90% of the write to iceberg in my current company is `overwrite` so it makes really hard/impossible to streaming read from these tables in Flink. I wonder if this patch requires Flink 1.13, or it'll also work for 1.12? 

@openinx do you know if this PR can make it to the next release? if so, roughly when will be the next release or does it depends on the progress on the priority 1 items in the roadmap?

**Comment by hameizi** [comment]
> I wonder if this patch requires Flink 1.13, or it'll also work for 1.12?

@xinbinhuang It work for both Flink 1.12 and 1.13.

**Comment by Initial-neko** [comment]
> > > > @MOBIN-F It's true, this PR translate update to -D and +I.
> > > 
> > > 
> > > It will query the record(1,qq). but this record is from before the update, Is that right?
> > 
> > 
> > If you don't query op of data, you will get only one result (1,aa).
> 
> Sorry, I don't understand how to get only one result(1,aa). Could you tell me how to do that?

https://ci.apache.org/projects/flink/flink-docs-release-1.13/zh/docs/dev/table/sqlclient/

client mode can help you to do that

**Comment by MOBIN-F** [comment]
> > > > > @MOBIN-F It's true, this PR translate update to -D and +I.
> > > > 
> > > > 
> > > > It will query the record(1,qq). but this record is from before the update, Is that right?
> > > 
> > > 
> > > If you don't query op of data, you will get only one result (1,aa).
> > 
> > 
> > Sorry, I don't understand how to get only one result(1,aa). Could you tell me how to do that?
> 
> https://ci.apache.org/projects/flink/flink-docs-release-1.13/zh/docs/dev/table/sqlclient/
> 
> client mode can help you to do that

Thanks ,After testing, the PR is problematic, there will be duplicate data in batch read, not suitable for Upsert stream and join scenario also have problems,
I'm  trying to implement this feature according to the design documentation

**Comment by hameizi** [comment]
> Thanks ,After testing, the PR is problematic, there will be duplicate data in batch read, not suitable for Upsert stream and join scenario also have problems,
> I'm trying to implement this feature according to the design documentation

This PR is not relate to bath mode. And can you descript the detail of Upsert and join problem?

**Comment by openinx** [comment]
@hameizi ,  would you mind to resolve the conflicts ?  I think I can take a look for the first round later.

**Comment by hameizi** [comment]
> @hameizi , would you mind to resolve the conflicts ? I think I can take a look for the first round later.

@openinx I will resolve the conflicts , but mayby take some time because there is many change in flink moudle. When do you have time to review? today?

**Comment by xianyouQ** [comment]
I merged this PR and  did some testing. i used flink cdc to consume binlog and wrote to iceberg table, then run some flink sql  query(streaming) and compared with mysql original query , then the results were not matched. 

After some debug I found 2 issues.

1. flink writing cdc data to iceberg will ignore UPDATE_BEFORE and treat UPDATE_AFTER as retract message,  it's completely not correct. The relevant codes are as follows:

`   

    //org.apache.iceberg.flink.sink.BaseDeltaTaskWriter
     switch (row.getRowKind()) {
      case INSERT:
      case UPDATE_AFTER:
        if (upsert) {
          writer.delete(row);
        }
        writer.write(row);
        break;

      case UPDATE_BEFORE:
        if (upsert) {
          break;  // UPDATE_BEFORE is not necessary for UPDATE, we do nothing to prevent delete one row twice
        }
        writer.delete(row);
        break;
      case DELETE:
        writer.delete(row);
        break;

      default:
        throw new UnsupportedOperationException("Unknown row kind: " + row.getRowKind());
    }
`


2.  when there is only delete operation during the period of one snapshot，the snapshot of iceberg will contains only equality delete files  and no data file(data file is deleted),  flink will ignore equality delete files and miss all  -D .   function org.apache.iceberg.ManifestGroup.planStreamingFiles()   create FileScanTask  only when data file exists, so flink can not process equality delete files.  


After fixed above 2 issues  the test is passed.

**Comment by hameizi** [comment]
> 1. flink writing cdc data to iceberg will ignore UPDATE_BEFORE and treat UPDATE_AFTER as retract message,  it's completely not correct. The relevant codes are as follows:

For this issue, config `write.upsert.enabled` is not suitable for CDC. So if your scene is CDC you can config it false that result `upsert` false.


> 2\. when there is only delete operation during the period of one snapshot，the snapshot of iceberg will contains only equality delete files  and no data file(data file is deleted),  flink will ignore equality delete files and miss all  -D .   function org.apache.iceberg.ManifestGroup.planStreamingFiles()   create FileScanTask  only when data file exists, so flink can not process equality delete files.

I will do more test for this issue.
Thanks for your feedback.

**Comment by xianyouQ** [comment]
> For this issue, config `write.upsert.enabled` is not suitable for CDC. So if your scene is CDC you can config it false that result `upsert` false.
Thanks , I also merged [change-delete-logic](https://github.com/hameizi/iceberg/tree/change-delete-logic)  , I think it's a better way to write delete rows.  I am confused why delete row writing to postition delete file also write to equality delete file.

**Comment by kingeasternsun** [comment]
> I merged this PR and did some testing. i used flink cdc to consume binlog and wrote to iceberg table, then run some flink sql query(streaming) and compared with mysql original query , then the results were not matched.
> 
> After some debug I found 2 issues.
> 
> 1. flink writing cdc data to iceberg will ignore UPDATE_BEFORE and treat UPDATE_AFTER as retract message,  it's completely not correct. The relevant codes are as follows:
> 
> `
> 
> ```
> //org.apache.iceberg.flink.sink.BaseDeltaTaskWriter
>  switch (row.getRowKind()) {
>   case INSERT:
>   case UPDATE_AFTER:
>     if (upsert) {
>       writer.delete(row);
>     }
>     writer.write(row);
>     break;
> 
>   case UPDATE_BEFORE:
>     if (upsert) {
>       break;  // UPDATE_BEFORE is not necessary for UPDATE, we do nothing to prevent delete one row twice
>     }
>     writer.delete(row);
>     break;
>   case DELETE:
>     writer.delete(row);
>     break;
> 
>   default:
>     throw new UnsupportedOperationException("Unknown row kind: " + row.getRowKind());
> }
> ```
> 
> `
> 
> 2. when there is only delete operation during the period of one snapshot，the snapshot of iceberg will contains only equality delete files  and no data file(data file is deleted),  flink will ignore equality delete files and miss all  -D .   function org.apache.iceberg.ManifestGroup.planStreamingFiles()   create FileScanTask  only when data file exists, so flink can not process equality delete files.
> 
> After fixed above 2 issues the test is passed.



> BaseDeltaTaskWriter

I think this code will be more clear if written like this
```java
  public void write(RowData row) throws IOException {
    RowDataDeltaWriter writer = route(row);

    switch (row.getRowKind()) {
      case INSERT:
        if (upsert) {
          writer.delete(row);
        }
        writer.write(row);
        break;
      case UPDATE_AFTER:
        writer.write(row);
        break;
      case UPDATE_BEFORE:
      case DELETE:
        writer.delete(row);
        break;

      default:
        throw new UnsupportedOperationException("Unknown row kind: " + row.getRowKind());
    }
  }
```

**Comment by liubo1022126** [comment]
@hameizi @openinx @rdblue hi，boys，We are currently using the flink upsert function. In which version is this PR planned to be released?

**Comment by xianyouQ** [comment]
Another issue, After  run rewrite data file action and expire snapshot action(retain one Last snapshot), and then started a new flink streaming job, it did not read the old data. I checked metadata files , there was only one snapshot and all manifest file had a smaller sequenceNumber. what should i do if i want to read all data.

File: core/src/main/java/org/apache/iceberg/ManifestGroup.java [hunk 1]
```diff
@@ -189,6 +189,48 @@ public CloseableIterable<FileScanTask> planFiles() {
     }
   }
 
+  /**
+   * Returns a iterable of scan tasks. It is safe to add entries of this iterable
+   * to a collection as {@link DataFile} in each {@link FileScanTask} is defensively
+   * copied.
+   * @return a {@link CloseableIterable} of {@link FileScanTask}
+   */
+  public CloseableIterable<FileScanTask> planStreamingFiles() {
```
Comment (stevenzwu) [inline core/src/main/java/org/apache/iceberg/ManifestGroup.java:198]: Streaming is an ambiguous name here. maybe `planFilesForChangelog`?

File: api/src/main/java/org/apache/iceberg/TableScan.java [hunk 2]
```diff
@@ -159,6 +159,17 @@ default TableScan select(String... columns) {
...
+
+  /**
+   * Create a new {@link TableScan} to read appended data for {@code snapshotId} exclusive to the current snapshot
+   * inclusive.
+   *
+   * @param snapshotId - the snapshot id read by the user, exclusive
+   * @return a table scan which can read append data for {@code snapshotId}
+   * exclusive and up to current snapshot inclusive
+   */
+  TableScan appendsCurrent(long snapshotId);
```
Comment (stevenzwu) [inline api/src/main/java/org/apache/iceberg/TableScan.java:171]: this change seems unnecessary

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/source/StreamingRowDataFileScanTaskReader.java [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+import org.apache.iceberg.orc.ORC;
+import org.apache.iceberg.parquet.Parquet;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.util.PartitionUtil;
+
+@Internal
+public class StreamingRowDataFileScanTaskReader implements FileScanTaskReader<RowData> {
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/source/StreamingRowDataFileScanTaskReader.java:50]: Similarly, maybe `ChangelogRowDataFileScanTaskReader` is more accurate

**Comment by hameizi** [comment]
@stevenzwu I will perfect this PR after https://github.com/apache/iceberg/pull/4580 finished, because there is some correlation between them.

**Comment by GavinH1984** [comment]
@hameizi hihi, is there any plan for this PR? i saw the PR [#4580](https://github.com/apache/iceberg/pull/4580) already get merge d into master,

**Comment by hameizi** [comment]
> @hameizi hihi, is there any plan for this PR? i saw the PR [#4580](https://github.com/apache/iceberg/pull/4580) already get merge d into master,

**Comment by hameizi** [comment]
> @hameizi hihi, is there any plan for this PR? i saw the PR [#4580](https://github.com/apache/iceberg/pull/4580) already get merge d into master,

I will work on this these days.

**Comment by iflytek-hmwang5** [comment]
> > @hameizi hihi, is there any plan for this PR? i saw the PR [#4580](https://github.com/apache/iceberg/pull/4580) already get merge d into master,
> 
> I will work on this these days.
[hameizi](https://github.com/hameizi) how about this PR?

**Comment by Bmyymwtao** [comment]
> > @hameizi hihi, is there any plan for this PR? i saw the PR [#4580](https://github.com/apache/iceberg/pull/4580) already get merge d into master,
> 
> I will work on this these days.

Hi, I pulled this PR and ran some tests，I found that when only deleteFile exists, the snapshot will not read the data. By viewing the ManifestGroup, I found that FileScanTask can only be generated when DataManifest exists.  Do you have the problem that only DeleteManifest has to consider reading -D data

**Comment by Xiangakun** [comment]
@hameizi @openinx Hello guys，what's the status of the issue now, please? Any blocking things? Thanks~

**Pull request closed without merging**